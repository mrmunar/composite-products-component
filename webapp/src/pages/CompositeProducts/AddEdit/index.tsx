import React, { useState, useEffect } from 'react';
import { useHistory } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';

import GenerateRecursiveComponentForm from '../../../components/compositeProducts/generateRecursiveComponentForm';
import { 
    addCompositeProducts, 
    getCompositeProduct, 
    editCompositeProduct
 } from '../../../redux/actions/compositeProducts';
import { validateCompositeProduct } from '../../../utils/validateCompositeProduct';
import ValidationMessage from '../../../components/base/form/validation/validationMessage';
import Button from '../../../components/base/form/button';

function AddEdit(props: any) {
    const dispatch = useDispatch();
    const history = useHistory();
    const loadStatus = useSelector((state: any) => {
        return state.compositeProducts
    });

    const compositeProductRecord = useSelector((state: any) => {
        return state.compositeProducts.record;
    });

    const [compositeProductState, setCompositeProductState] = useState({
        name: '',
        components: []
    });

    const [compositeProductName, setCompositeProductName] = useState('');
    const [isValid, setIsValid] = useState(true);

    const handleSetState = (state: any) => {
        setCompositeProductState(state);
        setIsValid(validateCompositeProduct(compositeProductState) && compositeProductName ? true : false);
    }

    const handleOnNameChange = (e: any) => {
        setCompositeProductName(e.target.value);
    }

    useEffect(() => {
        compositeProductState.name = compositeProductName;
        setIsValid(compositeProductName ? true : false);
    }, [compositeProductName, compositeProductState]);

    useEffect(() => {
        if (compositeProductRecord) {
            setCompositeProductState(compositeProductRecord);
            setCompositeProductName(compositeProductRecord.name);
        }
    }, [compositeProductRecord]);

    useEffect(() => {
        if (props.match.params.id) {
            dispatch(getCompositeProduct(props.match.params.id));
        }
    }, [dispatch, props]);
    
    const handleOnCancelClick = () => {
        history.push('/');
    }

    const handleOnSaveClick = () => {
        if (isValid) {
            if (!props.match.params.id) {
                dispatch(addCompositeProducts(compositeProductState));
            } else {
                dispatch(editCompositeProduct(props.match.params.id, compositeProductState));
            }
        }

        setTimeout(() => history.push('/'), 3000);
    }

    return (
        !loadStatus.isLoading ? 
        <div data-testid="add-edit-container">
            <div>Name</div>
            <input
                type="text"
                onChange={(e) => handleOnNameChange(e)}
                className="form-control"
                style={{ width: '50%' }}
                value={compositeProductName}
            />
            <GenerateRecursiveComponentForm
                onChange={(compositeProduct: any) => handleSetState(compositeProduct)}
                compositeProduct={compositeProductState} />
            <ValidationMessage valid={isValid} message="Please fill up all required fields" />
            <div className="text-right">
                <Button onClick={() => handleOnSaveClick()}label="Save" className="btn save-btn mr-2" />
                <Button onClick={() => handleOnCancelClick()} label="Cancel" className="btn cancel-btn" />
            </div>
        </div> : <div>Loading...</div>
    );
}

export default AddEdit;