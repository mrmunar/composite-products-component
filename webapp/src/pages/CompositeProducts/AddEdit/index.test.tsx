import React from 'react';
import { useSelector, useDispatch } from "react-redux";
import { render, cleanup } from '@testing-library/react';
import AddEdit from './';

afterEach(cleanup);

const testData = {
    id: 'efd0dab3-4813-44f0-98bc-ea3366582498',
    name: 'Example 1',
    components: [{
        type: 'PRODUCT',
        quantity: 1,
        productId: 'ba7a0995-4193-4589-8150-54a3cef5f2c6'
    }, {
        type: 'GROUP',
        label: 'Test Group',
        components: []
    }]
};

const invalidTestData = {
    id: 'efd0dab3-4813-44f0-98bc-ea3366582498',
    name: '',
    components: [{
        type: 'PRODUCT',
        quantity: 1,
        productId: 'ba7a0995-4193-4589-8150-54a3cef5f2c6'
    }, {
        type: 'GROUP',
        label: 'Test Group',
        components: []
    }]
};

jest.mock('react-redux', () => ({
    useDispatch: jest.fn(),
    useSelector: jest.fn()
}));

// Mocking GenerateRecursiveComponentForm component. Will test this in isolation
jest.mock('../../../components/compositeProducts/generateRecursiveComponentForm', () => () => 'GenerateRecursiveComponentForm');

const mockUseDispatch = useDispatch as jest.Mock;
const mockUseSelector = useSelector as jest.Mock;
const mockDispatch = jest.fn();

mockUseDispatch.mockImplementation(() => mockDispatch);
mockUseSelector.mockImplementation(state => testData);

test('should take a snapshot', () => {
    const { baseElement } = render(<AddEdit
        match={{ params: { id: 'some-test-id' } }}
    />);
    expect(baseElement).toMatchSnapshot();
});

test('validation message should appear if input data is invalid', () => {
    mockUseSelector.mockImplementation(state => invalidTestData);
    const { getByTestId } = render(<AddEdit
        onChange={() => { }}
        match={{ params: { id: 'some-test-id' } }}
    />);
    expect(getByTestId('add-edit-container')).toHaveTextContent('Please fill up all required fields');
});

test('validation message should be hidden if input data is valid', () => {
    mockUseSelector.mockImplementation(state => testData);

    const { getByTestId } = render(<AddEdit
        match={{ params: { id: 'some-test-id' } }}
    />);
    expect(getByTestId('add-edit-container')).not.toHaveTextContent('Please fill up all required fields');
});