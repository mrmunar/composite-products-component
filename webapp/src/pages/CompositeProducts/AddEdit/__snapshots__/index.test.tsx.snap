// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<body>
  <div>
    <div
      data-testid="add-edit-container"
    >
      <div>
        Name
      </div>
      <input
        class="form-control"
        style="width: 50%;"
        type="text"
        value="Example 1"
      />
      GenerateRecursiveComponentForm
      <div
        class="text-right"
      >
        <button
          class="btn save-btn mr-2"
          data-testid="button"
        >
          Save
        </button>
        <button
          class="btn cancel-btn"
          data-testid="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</body>
`;
